File: 06_black_swaption_pricing/python/black_swaption.py
import numpy as np

from scipy.stats import norm


class BlackSwaption:
    """
    Black lognormal swaption model.

    WARNING:
    --------
    Invalid when forward rates
    become negative.
    """

    def __init__(
        self,
        notional,
        strike,
        expiry,
        underlying_swap,
        volatility,
        payer=True
    ):

        self.notional = notional

        self.strike = strike

        self.expiry = expiry

        self.swap = underlying_swap

        self.volatility = volatility

        self.payer = payer

    def price(self, curve):

        F = self.swap.forward_swap_rate(
            curve
        )

        K = self.strike

        sigma = self.volatility

        T = self.expiry

        annuity = self.swap.annuity(
            curve
        )

        # ----------------------------------------
        # Black model invalid for negative rates
        # ----------------------------------------

        if F <= 0 or K <= 0:

            raise ValueError(
                "Black model invalid for "
                "negative rates."
            )

        d1 = (
            np.log(F / K)
            + 0.5 * sigma**2 * T
        ) / (sigma * np.sqrt(T))

        d2 = (
            d1
            - sigma * np.sqrt(T)
        )

        if self.payer:

            price = (
                self.notional
                * annuity
                * (
                    F * norm.cdf(d1)
                    - K * norm.cdf(d2)
                )
            )

        else:

            price = (
                self.notional
                * annuity
                * (
                    K * norm.cdf(-d2)
                    - F * norm.cdf(-d1)
                )
            )

        return price

File: 07_local_volatility_and_dupire/cpp/black_swaption.cpp
#include "black_swaption.hpp"

#include <cmath>

BlackSwaption::BlackSwaption(
    double N,
    double K,
    double T,
    double sigma,
    ForwardStartingSwap underlying_swap,
    bool is_payer
)
: swap(underlying_swap)
{

    notional = N;

    strike = K;

    expiry = T;

    volatility = sigma;

    payer = is_payer;
}

double BlackSwaption::normal_cdf(
    double x
){

    return 0.5 * erfc(-x/std::sqrt(2));
}

double BlackSwaption::price(
    YieldCurve& curve
){

    double F =
        swap.forward_swap_rate(curve);

    double K = strike;

    double sigma = volatility;

    double T = expiry;

    double annuity =
        swap.annuity(curve);

    double d1 =
        (std::log(F/K) + 0.5*sigma*sigma*T) / (sigma*std::sqrt(T));

    double d2 = d1 - sigma*std::sqrt(T);

    double value;

    if(payer){

        value =
            notional
            * annuity
            * (
                F*normal_cdf(d1)
                - K*normal_cdf(d2)
            );

    }else{

        value =
            notional
            * annuity
            * (
                K*normal_cdf(-d2)
                - F*normal_cdf(-d1)
            );
    }

    return value;
}
